### PR TITLE
create data channel sample

### PR DESCRIPTION
**Pull request by niyatim23**
Create a data channel sample
* Create a data channel from the viewer's side
* Communicate over the data channel by sending and receiving 
messages using `onopen` and `onnmessage` callbacks from master and viewer

### CODE REVIEW

File: samples/Common.c [hunk 1]
```diff
@@ -1,5 +1,6 @@
 #define LOG_CLASS "WebRtcSamples"
 #include "Samples.h"
+// #define MASTER_DATA_CHANNEL_MESSAGE "This message is from the KVS Master"
```
Comment (disa6302) [inline samples/Common.c:3]: Sugest moving this to `Sample.h` instead.

File: samples/Common.c [hunk 2]
```diff
@@ -21,6 +22,12 @@ VOID onDataChannelMessage(UINT64 customData, PRtcDataChannel pDataChannel, BOOL
     } else {
         DLOGI("DataChannel String Message: %.*s\n", pMessageLen, pMessage);
     }
+    // Send a response to the message sent by the viewer
```
Comment (disa6302) [inline samples/Common.c:25]: Is the customer expected to uncomment this to run? I am wondering if we can encapsulate all this into the `ENABLE_DATA_CHANNEL` directive which exists in the repository. That way we can avoid have commented code blocks.

File: samples/kvsWebRTCClientViewer.c [hunk 3]
```diff
@@ -165,6 +195,26 @@ INT32 main(INT32 argc, CHAR* argv[])
         goto CleanUp;
     }
 
+    PRtcDataChannel pDataChannel = NULL;
```
Comment (disa6302) [inline samples/kvsWebRTCClientViewer.c:198]: Encapsulate this entire section this way:

```
#ifdef ENABLE_DATA_CHANNEL
<code block>
#endif

Checkout usage [here](https://github.com/awslabs/amazon-kinesis-video-streams-webrtc-sdk-c/blob/c4a13309b6559d19529ca8049a50d4e2b241564f/src/source/PeerConnection/PeerConnection.c#L143)

File: samples/kvsWebRTCClientViewer.c [hunk 4]
```diff
@@ -165,6 +195,26 @@ INT32 main(INT32 argc, CHAR* argv[])
...
+    printf("[KVS Viewer] Creating data channel...completed\n");
+
+    // Setting a callback for when the data channel is open
+    retStatus = dataChannelOnOpen(pDataChannel, (UINT64) &datachannelLocalOpenCount, dataChannelOnOpenCallback);
+    if(retStatus != STATUS_SUCCESS) {
+        printf("[KVS Viewer] dataChannelOnOpen(): operation returned status code: 0x%08x \n", retStatus);
+        goto CleanUp;
+    }
+    printf("[KVS Viewer] Data Channel open now...\n");
+
```
Comment (disa6302) [inline samples/kvsWebRTCClientViewer.c:217]: We should look into cleaning up data channels at the application exit for which we require `dataChannelClose()` API. We can handle that in a new PR.

File: samples/Common.c (hunk 1 above)
Comment (niyatim23) [reply samples/Common.c]: Moved #define MASTER_DATA_CHANNEL_MESSAGE and #define VIEWER_DATA_CHANNEL_MESSAGE to Samples.h

File: samples/Common.c (hunk 2 above)
Comment (niyatim23) [reply samples/Common.c:24]: Added `ENABLE_DATA_CHANNEL` in necessary places

File: samples/kvsWebRTCClientViewer.c (hunk 3 above)
Comment (niyatim23) [reply samples/kvsWebRTCClientViewer.c:192]: Added ENABLE_DATA_CHANNEL encapsulation here

File: samples/kvsWebRTCClientViewer.c [hunk 5]
```diff
@@ -2,6 +2,31 @@
 
 extern PSampleConfiguration gSampleConfiguration;
 
+// onMessage callback for a message received by the viewer on a data channel
+VOID dataChannelOnMessageCallback(UINT64 customData, PRtcDataChannel pDataChannel, BOOL isBinary, PBYTE pMessage, UINT32 pMessageLen)
+{
+    UNUSED_PARAM(customData);
+    STATUS retStatus = STATUS_SUCCESS;
```
Comment (hassanctech) [inline samples/kvsWebRTCClientViewer.c:9]: `retStatus` is unused so it can be removed.

Comment (hassanctech) [reply samples/kvsWebRTCClientViewer.c:9]: Looks like `pDataChannel` is unused as well.  If this is required to match the signature (I didn't check) then you can just mark it as `UNUSED_PARAM()`

File: samples/kvsWebRTCClientViewer.c [hunk 6]
```diff
@@ -2,6 +2,31 @@
...
+        DLOGI("DataChannel String Message: %.*s\n", pMessageLen, pMessage);
+    }
+}
+
+// onOpen callback for the onOpen event of a viewer created data channel
+VOID dataChannelOnOpenCallback(UINT64 customData, PRtcDataChannel pDataChannel) {
+    DLOGI("New DataChannel has been opened %s \n", pDataChannel->name);
+    dataChannelOnMessage(pDataChannel, customData, dataChannelOnMessageCallback);
+    ATOMIC_INCREMENT((PSIZE_T) customData);
+    STATUS retStatus = STATUS_SUCCESS;
```
Comment (hassanctech) [inline samples/kvsWebRTCClientViewer.c:22]: To adhere with C89 standard variables need to be declared the top of a scope block.  You will need to move the declaration of `retStatus` to the top of this function.

**Comment by hassanctech** [comment]
Looks great overall, some minor tweaks necessary to pass CI.

File: samples/kvsWebRTCClientViewer.c (hunk 6 above)
Comment (niyatim23) [reply samples/kvsWebRTCClientViewer.c]: Moved the declaration of `retStatus` to the top of this function.

File: samples/kvsWebRTCClientViewer.c (hunk 5 above)
Comment (niyatim23) [reply samples/kvsWebRTCClientViewer.c]: Marked `pDataChannel` as `UNUSED_PARAM()`. Removed `retStatus`.

**Review by disa6302 — APPROVED** [review]

**Pull request merged**